util/mdbook/move_md.py

```diff
@@ -0,0 +1,142 @@
+logger = logging.getLogger(__name__)
+
+logging.getLogger('markdown_it').setLevel(logging.WARNING)
+
+
+# https://stackoverflow.com/a/71874881
+def relative(target: Path, origin: Path) -> Path:
+    """Return the path of target relative to origin.
+    i.e. --> pathlib equivalent of os.path.relpath'
+    """
+    try:
+        return Path(target).resolve().relative_to(Path(origin).resolve())
+    except ValueError:  # target does not start with origin
+        # recursion with origin (eventually origin is root so try will succeed)
+        return Path('..').joinpath(relative(target, Path(origin).parent))
+
+
+def do_update(l: str, old_path: pathlib.Path) -> bool:
+    """Do some checking on the relative link in the existing file.
+
+    Return 'False' if we don't need to recalculate it.
+    """
+
+    # Don't update link if it is a URL
+    if validators.url(l):
+        logger.debug(f"IsUrl : {l}")
+        return False
+
+    # Strip fragment/anchor if present
+    l_path, l_frag = None, None
+    if "#" in l:
+        l_path, l_frag = l.split("#")
+    else:
+        l_path, l_frag = l, None
+    if not l_path:
+        # Must be a local anchor in the same file
+        # We don't need to update this one.
+        return False
+
+    oldlink_target = (old_path.parents[0]/Path(l_path)).resolve()
+    if not oldlink_target.exists():
+        logger.warning("Link in old_path points to non-existent file.")
+        logger.warning(f"  OLD_PATH       : {old_path}")
+        logger.warning(f"  REL_LINK       : {l}")
+        logger.warning(f"  TARGET   (BAD) : {oldlink_target}")
+
+    return True
+
+
+def move_md(old_path: pathlib.Path, new_path: pathlib.Path) -> None:
+    '''Main link rewriter'''
+    if new_path.suffix != '.md':
+        logger.error(f"new_path does not end in .md : {new_path}")
+    if old_path.suffix != '.md':
+        logger.error(f"old_path does not end in .md : {old_path}")
+    if not old_path.exists():
+        logger.error(f"old_path does not exist! : {old_path}")
+
+    rel = relative(old_path.parent, new_path.parent)
+    logger.info(f"OLD path : {old_path}")
+    logger.info(f"NEW path : {new_path}")
+    logger.info(f"relative path to recalculate links from NEW to OLD : {rel}")
+
+    old_markdown = old_path.read_text()
+    new_md = old_markdown  # Copy old file, then modify file in-memory
+
+    # Extract all links from the old markdown file
+    ast = SyntaxTreeNode(MarkdownIt().parse(old_markdown))
+    # Ensure list is in-order
+    links = []
+    for t in ast.walk():
+        if getattr(t, 'type') in ('link'):
+            links.append(t.attrs['href'])
+        if getattr(t, 'type') in ('image'):
+            links.append(t.attrs['src'])
+
+    # Create new markdown with all relative links recalculated
+    idx = 0
+    for l in links:
+        if not do_update(l, old_path):
+            continue
+        idx = new_md.find(l, idx)
```

Comment:
I would be a bit worried about such a global replacement. If the link is something trivial like "." or ".." or even "index", you might replace the wrong things in the file. Since you use MarkdownIt to parse it, can it give you the offsets in the string?

Reply:
So the AST did not contain span information when I checked, hence the slightly awkward 'assembled ordered list of matches, then find and replace forward storing the index of the find each time'. It would be nice just to modify the AST then re-render the file with a markdown renderer, but trying this out I would always see other unrelated changes to the formatting which is undesirable.
If you have any suggestions about quick and easy ways to make this more robust I'm all ears, but I also don't wan't to prematurely-optimize here. Since moving/splitting markdown files will only happen manually with committed changes to the repo, it was easy to go through the result and check for malformed links etc. so I just did a sanity check on the diff.

Reply:
Looking at [https://markdown-it-py.readthedocs.io/en/latest/api/markdown_it.tree.html?highlight=syntaxtree#markdown_it.tree.SyntaxTreeNode], I see a `map` property that gives source line span only, it's better than nothing but not quite accurate enough. I do agree though that changes will be reviewed manually anyway so it's probably not a big deal.

Reply:
Unfortunately, the map field is not populated for links.
e.g.

```python
     for t in ast.walk():
         if getattr(t, 'type') in ('link'):
             links.append(t.attrs['href'])
+            logger.info(f"map = {t.map}")
         if getattr(t, 'type') in ('image'): 
             links.append(t.attrs['src'])
+            logger.info(f"map = {t.map}")
```

```sh
harry@hcallahan-neptune:~/projects/opentitan$ ./util/mdbook/move_md.py -O ./doc/introduction.md -N ./doc/doc2/intro.md --verbose
INFO:__main__:OLD path : /home/harry/projects/opentitan/doc/introduction.md
INFO:__main__:NEW path : /home/harry/projects/opentitan/doc/doc2/intro.md
INFO:__main__:relative path to recalculate links from NEW to OLD : ..
INFO:__main__:map = None
INFO:__main__:map = None
INFO:__main__:map = None
INFO:__main__:map = None
INFO:__main__:map = None
INFO:__main__:map = None
INFO:__main__:map = None
INFO:__main__:map = None
INFO:__main__:map = None
INFO:__main__:map = None
INFO:__main__:map = None
INFO:__main__:map = None
INFO:__main__:map = None
INFO:__main__:map = None
INFO:__main__:map = None
INFO:__main__:map = None
INFO:__main__:map = None
INFO:__main__:map = None
INFO:__main__:map = None
INFO:__main__:map = None
(...)
```

Would be nice to do something better in the future.